grakn-engine/src/main/java/ai/grakn/engine/postprocessing/PostProcessor.java

```diff
@@ -0,0 +1,274 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import redis.clients.jedis.Jedis;
+import redis.clients.util.Pool;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.locks.Lock;
+import java.util.stream.Collectors;
+
+import static com.codahale.metrics.MetricRegistry.name;
+
+/**
+ * <p>
+ *     Class responsible for post processing concepts which may be broken.
+ * </p>
+ *
+ * <p>
+ *     Contains the logic which is needed for post processing {@link ai.grakn.concept.Concept}s for the moment this
+ *     includes:
+ *
+ *     - Merging duplicate {@link ai.grakn.concept.Attribute}s
+ *     - Counting new {@link ai.grakn.concept.Thing}s which have been created under {@link ai.grakn.concept.Type}s
+ * </p>
+ *
+ * @author Filipe Peliz Pinto Teixeira
+ */
+public class PostProcessor {
+    private final static Logger LOG = LoggerFactory.getLogger(PostProcessor.class);
+    private final GraknEngineConfig engineConfig;
+    private final MetricRegistry metricRegistry;
+    private final LockProvider lockProvider;
+    private final RedisCountStorage redis;
+    private final EngineGraknTxFactory factory;
+
+    @Deprecated
+    private static final String LOCK_KEY = "/post-processing-lock";
+
+    private PostProcessor(GraknEngineConfig engineConfig, Pool<Jedis> jedisPool, EngineGraknTxFactory factory, LockProvider lockProvider, MetricRegistry metricRegistry){
+        this.engineConfig = engineConfig;
+        this.metricRegistry = metricRegistry;
+        this.lockProvider = lockProvider;
+        this.redis = RedisCountStorage.create(jedisPool, metricRegistry);
+        this.factory = factory;
+    }
+
+    public static PostProcessor create(GraknEngineConfig engineConfig, Pool<Jedis> jedisPool, EngineGraknTxFactory factory, LockProvider lockProvider, MetricRegistry metricRegistry){
+        return new PostProcessor(engineConfig, jedisPool, factory, lockProvider, metricRegistry);
+    }
+
+    /**
+     * Updates the counts of {@link ai.grakn.concept.Type}s based on the commit logs received.
+     *
+     * @param keyspace The {@link Keyspace} which contains {@link ai.grakn.concept.Type}s with new instances.
+     * @param commitLog The commit log containing the details of the job
+     */
+    public void updateCounts(Keyspace keyspace, Json commitLog){
+        final long shardingThreshold = engineConfig.getProperty(GraknConfigKey.SHARDING_THRESHOLD);
+        final int maxRetry = engineConfig.getProperty(GraknConfigKey.LOADER_REPEAT_COMMITS);
+        try (Timer.Context context = metricRegistry.timer(name(PostProcessor.class, "execution")).time()) {
+            Map<ConceptId, Long> jobs = getCountUpdatingJobs(commitLog);
+            metricRegistry.histogram(name(PostProcessor.class, "jobs"))
+                    .update(jobs.size());
+
+            //We Use redis to keep track of counts in order to ensure sharding happens in a centralised manner.
```

Comment:
If the problem here is only the fact that each engine has a different snapshot, this is something we can solve with a G-Counter 
https://github.com/aphyr/meangirls#g-counter

Reply:
(without redis)

Reply:
Also, just a comment, not for this PR

Reply:
Oh that is cool, but sadly we need to be able to decrement as well. For example when you delete instances.

Reply:
There's the PN-counter for that. The downside of course is that instead of a number we are saving (number of engines * 2) numbers in that case. It might still be good enough to avoid synchronisation and Redis